ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncActions.java

```diff
@@ -126,6 +131,19 @@ private void waitForAnyPeer(final CompletableFuture<Void> result) {
   }
 
   public CompletableFuture<FastSyncState> selectPivotBlock() {
+    return loadPivotBlockFromStorage().orElseGet(this::selectPivotBlockFromPeers);
```

Comment:
(optional / discussion) What about loading this from storage in the constructor?  Otherwise, successive calls to this method may always return the same value, which you may not want or may be unexpected.  Also thinking it might be nice to move the reading / writing of this to the `FastSyncState` class with something like `FastSyncState.create(dataDir)` that loads or creates a new state.

Reply:
yeah I think there are definitely some improvements along these lines. The changes have wound up being reasonable large though so I'm going to merge this and put up a follow-up PR so we can review them and discuss more clearly.